feat: new storybook and sveltekit blog post (#2244)
mainmatter/mainmatter.com: src/posts/2023-11-27-sveltekit-storybook.md

---
title: "SvelteKit 🤝 Storybook"
authorHandle: paoloricciuti

> tags: [svelte, sveltekit, storybook]

Reviewer: ```suggestion
tags: [svelte, storybook]
```

I'd say let's collect all Svelte posts under the `svelte` tag – otherwise we have too many tags with too few posts each

bio: "Paolo Ricciuti (paoloricciuti)"
description:
  "Paolo Ricciuti (paoloricciuti) shows the newest addition to Storybook that allows for a tighter integration between Storybook and SvelteKit"
og:
  image: /assets/images/posts/2023-11-27-sveltekit-storybook/og-image.jpg

> tagline: |

Reviewer: This is the first paragraph of the post actually – e.g. _"There are numerous OAuth articles…"_ on [this one](https://mainmatter.com/blog/2023/11/23/setting-up-oauth-with-auth-js-and-sveltekit/).

Author: Oh ok fixing it now

    sveltekit_experimental: {
      stores: {
        page: {}, // this mocks the page store
        navigating: {}, // this mocks the navigating store
        update: true, // this mocks the update store
      },
      navigation: {
        goto: () => {}, // this is a callback for the goto method
        invalidate: () => {}, // this is a callback for the invalidate method
        invalidateAll: () => {}, // this is a callback for the invalidateAll method
        afterNavigate: {}, // this will be the object passed to `afterNavigate` on mount
      },
      forms: {
        enhance: () => {}, // this is a callback for enhance
      },
      hrefs: {
        "/test": {
          callback: () => {}, // this is a called when a link to /test is clicked
        },
        "/test/.+": {
          callback: () => {}, // this is a called when a link that match /test/.+ is clicked
          asRegex: true,
        },
      },
    },
  },
};
```

well...this was a lot of information packed into a small snippet of code. The
gist of it is that you have three properties, one for every module exported from
`$app` plus an `href` property. So, for example, to customize the behavior of

> the `page` store exported from `$app/stores` you can set

Reviewer 2: ```suggestion
the `page` store exported from `$app/stores`, you can set
```

> `sveltekit_experimental.stores.page` or to customize the behavior of the `goto`

Reviewer 2: ```suggestion
`sveltekit_experimental.stores.page`, or to customize the behavior of the `goto`
```

> function exported from `$app/navigation` you can set

Reviewer 2: ```suggestion
function exported from `$app/navigation`, you can set
```

`sveltekit_experimental.navigation.goto`.

Let's see what you can customize.

### `stores.page`

The object you assign to this property will be the value of the `page` store.
You are free to assign only a partial (what you need in the story)

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      stores: {
        page: {
          data: {
            locale: "en",
          },
        },
      },
    },
  },
};
```

### `stores.navigating`

The object you assign to this property will be the value of the `navigating`

> store. You are free to assign only a partial (what you need in the story)

Reviewer 2: ```suggestion
store. You are free to assign only a partial (what you need in the story).
```

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      stores: {
        navigating: {
          to: {
            route: {
              id: "/test-route/[id]",
            },
          },
        },
      },
    },
  },
};
```

### `stores.update`

The object you assign to this property will be the value of the `update` store.
The value of the store is a boolean.

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      stores: {
        update: false,
      },
    },
  },
};
```

### `navigation.goto`

The function you assign to this property will be called whenever the Story you
are rendering calls `goto`. You will receive the same parameters passed to the

> `goto` function by the component. If you don't pass any function to this

Reviewer 2: ```suggestion
`goto` function by the component. If you don't pass any function to this,
```

Storybook will log an action in the actions panel.

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      navigation: {
        goto: (to, options) => {
          console.log("The user tried to navigate to", to);
        },
      },
    },
  },
};
```

### `navigation.invalidate`

The function you assign to this property will be called whenever the Story you
are rendering calls `invalidate`. You will receive the same parameters passed to
the `invalidate` function by the component. If you don't pass any function to

> this Storybook will log an action in the actions panel.

Reviewer 2: ```suggestion
this, Storybook will log an action in the actions panel.
```

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      navigation: {
        invalidate: (toInvalidate, options) => {
          console.log("The user invalidated", toInvalidate);
        },
      },
    },
  },
};
```

### `navigation.invalidateAll`

The function you assign to this property will be called whenever the Story you

> are rendering calls `invalidateAll`. If you don't pass any function to this

Reviewer 2: ```suggestion
are rendering calls `invalidateAll`. If you don't pass any function to this,
```

Storybook will log an action in the actions panel.

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      navigation: {
        invalidateAll: () => {
          console.log("The user called invalidatedAll");
        },
      },
    },
  },
};
```

### `navigation.afterNavigate`

Generally, `afterNavigate` is called by SvelteKit after each navigation. This
doesn't really make sense in Storybook since there's no router involved.
Instead, if you call `afterNavigate` in your component, the function you passed
as a parameter will be called on mount. The function you pass takes the
information about the navigation as an argument and you can specify that

> argument using `navigation.afterNavigate` (you can pass a partial)

Reviewer 2: ```suggestion
argument using `navigation.afterNavigate` (you may also pass a partial).
```

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      navigation: {
        afterNavigate: {
          from: {
            route: {
              id: "/",
            },
          },
        },
      },
    },
  },
};
```

### `forms.enhance`

> If you are using the `enhance` action on a form when inside Storybook the form

Reviewer 2: ```suggestion
If you are using the `enhance` action on a form when inside Storybook, the form
```

submission will still be prevented and the function you assign to
`forms.enhance` will be called.

```ts
export const Primary: Story = {
  args: {
    primary: true,
    label: "Button",
  },
  parameters: {
    sveltekit_experimental: {
      forms: {
        enhance: () => {
          console.log("The user submitted a form");
        },
      },
    },
  },
};
```

### `hrefs`

> By default every link will log an action to the Actions panel but if you want

Reviewer 2: ```suggestion
By default, every link will log an action to the Actions panel, but if you want
```

you can pass another property to the `sveltekit_experimental` parameter. `hrefs`
is an object where every key is a string that should represent a link inside
your Story, and the value is typed as such:
`{ callback: () => void, asRegex?: boolean }`. As you might have guessed, the
`callback` function will be called whenever the user clicks on an `A` tag that

> has an `href` equal to the key. If you specified `asRegex` as true the function

Reviewer 2: ```suggestion
has a `href` equal to the key. If you specified `asRegex` as true, the function
```

will be called whenever your regular expression matches the `href` attribute of
the tag.

## Conclusions

> And that was it: this was a brief introduction to those new features and I'm

Reviewer 2: ```suggestion
And that was it: this was a brief introduction to the new features, and I'm
```

sure you will find crazy ways to use those functionalities. I'm so happy to have

> been able to contribute to such a pivotal project (also i want to thank Jeppe

Reviewer 2: ```suggestion
been able to contribute to such a pivotal project, and more so to be able to use those features myself! Also, I want to thank Jeppe
```

> for helping me become familiar with this huge codebase) but i'm more than happy

Reviewer 2: ```suggestion
for helping me become familiar with this huge codebase.
```

> that i will soon be able to use those features myself!

Reviewer 2: ```suggestion
```
